tools/build-tools/src/test/typeValidator/interfaceValidator.spec.ts

```diff
@@ -0,0 +1,391 @@
+        const oldTypeData = getTypeDataForSource(oldSourceFile);
+        const newSourceFile = project.createSourceFile(`${pkgDir}/src/interfaceNew.ts`, newSource, { overwrite: true });
+        const newTypeData = getTypeDataForSource(newSourceFile);
+
+        const validator = createSpecificValidator(
+            project.getTypeChecker(),
+            oldTypeData.node,
+            project.getTypeChecker(),
+            newTypeData.node,
+        );
+
+        return validator.validate(project, pkgDir);
+    }
+
+    describe("methods", () => {
+        // scenario: added new method
+        // expected result: major breaking change
+        it("adds a new required method declaration", () => {
+            const sourceOld =
+            `
+            export interface ITestInterface {}
+            `;
+
+            const sourceNew =
+            `
+            export interface ITestInterface {
+                newMethod();
+            }
+            `;
+
+            let increment = checkIncrement(sourceOld, sourceNew);
+            assert.strictEqual(increment, BreakingIncrement.major);
+        });
+
+        // adding an optional method is an incremental change
+        it("adds a new optional method declaration", () => {
+            const sourceOld =
+            `
+            export interface ITestInterface {}
+            `;
+
+            const sourceNew =
+            `
+            export interface ITestInterface {
+                newMethod?();
+            }
+            `;
+
+            let increment = checkIncrement(sourceOld, sourceNew);
+            assert.strictEqual(increment, BreakingIncrement.minor);
+        });
+
+        // removing a method is a breaking change
+        it("removed a method declaration", () => {
+            const sourceOld =
+            `
+            export interface ITestInterface {
+                oldMethod();
+            }
+            `;
+
+            const sourceNew =
+            `
+            export interface ITestInterface {}
+            `;
+
+            let increment = checkIncrement(sourceOld, sourceNew);
+            assert.strictEqual(increment, BreakingIncrement.major);
+        });
+
+        // making a required method optional is a breaking change
+        it("adds a new optional method declaration", () => {
+            const sourceOld =
+            `
+            export interface ITestInterface {
+                newMethod();
+            }
+            `;
+
+            const sourceNew =
+            `
+            export interface ITestInterface {
+                newMethod?();
+            }
+            `;
+
+            let increment = checkIncrement(sourceOld, sourceNew);
+            assert.strictEqual(increment, BreakingIncrement.major);
+        });
+
+        // getters are treated the same as public methods
+        // adding/removing one is a breaking change
+        // TODO: getters currently unsupported in ts-morph but we don't use this/it's rare
```

Comment:
Let's add an eslint rule using [no-restricted-syntax](https://eslint.org/docs/rules/no-restricted-syntax#no-restricted-syntax). That will help prevent new introductions.

Reply:
Oh... this is interfaces, so it's not JS... nvm